File: typescript/dnd-character/dnd-character.ts
export class DnDCharacter {
  private hitpoints: number = 10
  private strength: number = DnDCharacter.generateAbilityScore() // 强度
  private dexterity: number = DnDCharacter.generateAbilityScore() // 敏捷
  private constitution: number = DnDCharacter.generateAbilityScore() //体质
  private intelligence: number = DnDCharacter.generateAbilityScore() // 智力
  private wisdom: number = DnDCharacter.generateAbilityScore() // 智慧
  private charisma: number = DnDCharacter.generateAbilityScore() // 魅力

  constructor() {
    this.hitpoints = this.hitpoints + DnDCharacter.getModifierFor(this.constitution)
  }

  public static generateAbilityScore(): number {
    let numbers: Array<number> = []
    for (let i = 0; i < 3; i++) {
      numbers.push(Math.floor(Math.random() * 6) + 1)
    }
    numbers.sort((a, b) => a - b).shift()
    return numbers.reduce((acc, curr) => acc += curr, 0)
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2)
  }
}

File: zig/dnd-character/dnd_character.zig
const std = @import("std");

var rand = std.rand.DefaultPrng.init(0);

pub fn modifier(score: i8) i8 {
    return @divFloor(score - 10, 2);
}

pub fn ability() i8 {
    var total: i8 = 0;
    var min: i8 = 6;
    for (0..4) |_| {
        const r = rand.random().intRangeAtMost(i8, 1, 6);
        total += r;
        if (r < min) {
            min = r;
        }
    }
    return total - min;
}

pub const Character = struct {
    strength: i8,
    dexterity: i8,
    constitution: i8,
    intelligence: i8,
    wisdom: i8,
    charisma: i8,
    hitpoints: i8,

    pub fn init() Character {
        const constitution = ability();
        return Character{ .strength = ability(), .dexterity = ability(), .constitution = constitution, .intelligence = ability(), .wisdom = ability(), .charisma = ability(), .hitpoints = 10 + modifier(constitution) };
    }
};
